src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs

```diff
@@ -1102,5 +1104,218 @@ public void FormattingReturnTypeOptions()
+                         ptr.ToTestDisplayString());
+        }
+
+        [Fact]
+        public void PublicApi_CreateOutTypeSymbol()
+        {
+            var comp = (Compilation)CreateCompilation("");
+            var @string = comp.GetSpecialType(SpecialType.System_String);
+            var @void = comp.GetSpecialType(SpecialType.System_Void);
+
+            var ptr = comp.CreateFunctionPointerTypeSymbol(
+                @void,
+                RefKind.None,
+                ImmutableArray.Create((ITypeSymbol)@string),
+                ImmutableArray.Create(RefKind.Out));
+
+            Assert.Equal("delegate*<out modreq(System.Runtime.InteropServices.OutAttribute) System.String, System.Void>",
+                         ptr.ToTestDisplayString());
+        }
+
+        [Fact]
+        public void PublicApi_InOutAttributeMissing()
+        {
+            var comp = (Compilation)CreateCompilation("");
+            comp.MakeTypeMissing(WellKnownType.System_Runtime_InteropServices_InAttribute);
+            comp.MakeTypeMissing(WellKnownType.System_Runtime_InteropServices_OutAttribute);
+            var @string = comp.GetSpecialType(SpecialType.System_String);
+
+            var ptr = comp.CreateFunctionPointerTypeSymbol(
+                @string,
+                RefKind.RefReadOnly,
+                ImmutableArray.Create((ITypeSymbol)@string),
+                ImmutableArray.Create(RefKind.Out));
+
+            Assert.Equal("System.Runtime.InteropServices.InAttribute[missing]", ptr.Signature.RefCustomModifiers.Single().Modifier.ToTestDisplayString());
+            Assert.Equal("System.Runtime.InteropServices.OutAttribute[missing]", ptr.Signature.Parameters.Single().RefCustomModifiers.Single().Modifier.ToTestDisplayString());
+        }
+
+        [Fact]
+        public void PublicApi_GetTypeInfo()
+        {
+            var comp = CreateFunctionPointerCompilation(@"
+unsafe class C
+{
+    public static string M1(C c) => null;
+    public static void M2(string s, int it) {}
+    public delegate*<string, int, void> M(delegate*<C, string> ptr)
+    {
+        delegate*<string, int, void> ptr2 = &M2;
+        ptr = &M1;
+        ptr(null);
+        return &M2;
+    }
+}");
+
+            comp.VerifyDiagnostics();
+
+            var syntaxTree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(syntaxTree);
+
+            var mDeclSyntax = syntaxTree.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().Skip(2).Single();
+
+            verifyVoidReturn(mDeclSyntax.ReturnType, useConvertedType: false);
+            verifyStringReturn(mDeclSyntax.ParameterList.Parameters[0].Type!, useConvertedType: false);
+
+            var varDecl = mDeclSyntax.DescendantNodes().OfType<VariableDeclarationSyntax>().Single();
+            verifyVoidReturn(varDecl.Type, useConvertedType: false);
+            verifyVoidReturn(varDecl.Variables.Single().Initializer!.Value, useConvertedType: true);
```

Comment:
>varDecl.Variables.Single().Initializer!.Value [](start = 29, length = 45)

I think we should also assert SemanticModel behavior for the operand of an address of. I am assuming we are testing an address of node here.

Reply:
`VerifyFunctionPointerSemanticInfo` will do this, asserting that it's the same expected symbols as the containing address of expression.

---
In reply to: [428913130](https://github.com/dotnet/roslyn/pull/44436#discussion_r428913130) [](ancestors = 428913130)